src/compiler/transformers/esDecorators.ts

```diff
@@ -0,0 +1,2087 @@
+            //           i. Set _v_ to ? NamedEvaluation of |Initializer| with argument _bindingId_.
+            //     ...
+
+            if (!node.dotDotDotToken &&
+                isIdentifier(node.name) && node.initializer &&
+                isDecoratedAnonymousClassExpression(node.initializer)) {
+                const assignedName = getAssignedNameOfIdentifier(node.name, node.initializer);
+                const propertyName = visitNode(node.propertyName, visitor, isPropertyName);
+                const name = visitNode(node.name, visitor, isBindingName);
+                const initializer = visitNode(node.initializer, node => namedEvaluationVisitor(node, assignedName), isExpression);
+                return factory.updateBindingElement(node, /*dotDotDotToken*/ undefined, propertyName, name, initializer);
+            }
+
+            return visitEachChild(node, visitor, context);
+        }
+
+        function visitDestructuringAssignmentTarget(node: LeftHandSideExpression): Expression {
+            if (isObjectLiteralExpression(node) || isArrayLiteralExpression(node)) {
+                return visitAssignmentPattern(node);
+            }
+
+            if (isSuperProperty(node) && classThis && classSuper) {
+                const propertyName =
+                    isElementAccessExpression(node) ? visitNode(node.argumentExpression, visitor, isExpression) :
+                    isIdentifier(node.name) ? factory.createStringLiteralFromNode(node.name) :
+                    undefined;
+                if (propertyName) {
+                    const paramName = factory.createTempVariable(/*recordTempVariable*/ undefined);
+                    const expression = factory.createAssignmentTargetWrapper(
+                        paramName,
+                        factory.createReflectSetCall(
+                            classSuper,
+                            propertyName,
+                            paramName,
+                            classThis,
+                        )
+                    );
+                    setOriginalNode(expression, node);
+                    setTextRange(expression, node);
+                    return expression;
+                }
+            }
+
+            return visitEachChild(node, visitor, context);
+        }
+
+        function visitAssignmentElement(node: Exclude<ArrayAssignmentElement, SpreadElement | OmittedExpression>) {
+            // 13.15.5.5 RS: IteratorDestructuringAssignmentEvaluation
+            //   AssignmentElement : DestructuringAssignmentTarget Initializer?
+            //     ...
+            //     4. If |Initializer| is present and _value_ is *undefined*, then
+            //        a. If IsAnonymousFunctionDefinition(|Initializer|) and IsIdentifierRef of |DestructuringAssignmentTarget| are both *true*, then
+            //           i. Let _v_ be ? NamedEvaluation of |Initializer| with argument _lref_.[[ReferencedName]].

```

Comment:
NamedEvaluation has been in the spec for quite a while, right? Have we ignored it until now?

Reply:
We have ignored it until now, since there wasn't an extremely strong push for it. However, the `name` is fairly important to class decorators since it's one of the few things actually stored on the `context` object.

I may retrofit `__setFunctionName` into other transforms, but want to try to address the complexity around visiting and working with assignment patterns before tackling this anywhere else.